python/llm/src/bigdl/llm/transformers/models/yuan.py

```diff
@@ -62,46 +91,95 @@ def yuan_attention_forward(
+                                                                         self.head_dim)
+        kv_seq_len += 1
 
-        query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
-        key_states = key_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
     else:
-        hidden_states = self.lf_gate(hidden_states, before_hidden_states)
-        query_states = self.q_proj(hidden_states)
-        key_states = self.k_proj(hidden_states)
-        qk_states = torch.cat([query_states, key_states], dim=-1)
-        qk_states = qk_states.view(bsz, q_len,
-                                   self.num_heads,
-                                   int(qk_states.shape[-1]//self.num_heads))
-        (query_states, key_states) = torch.chunk(qk_states, 2, dim=-1)
-        query_states = query_states.transpose(1, 2)
-        key_states = key_states.transpose(1, 2)
-
-    kv_seq_len = key_states.shape[-2]
-    if past_key_value is not None:
-        kv_seq_len += past_key_value[0].shape[-2]
-    cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
-    query_states, key_states = apply_rotary_pos_emb(query_states,
-                                                    key_states,
-                                                    cos, sin,
-                                                    position_ids,
-                                                    "yuan")
-
-    if past_key_value is not None:
-        # reuse k, v, self_attention
-        key_states = torch.cat([past_key_value[0], key_states], dim=2)
-        value_states = torch.cat([past_key_value[1], value_states], dim=2)
+        value_states = \
+            self.v_proj(hidden_states).view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
+        if self.use_shareqk:
+            # use_shareqk is disabled for now
+            qk_states = self.qk_proj(hidden_states).view(bsz, q_len, self.num_heads*self.head_dim)
+            query_key = qk_states.unsqueeze(2) * self.qk_weight + self.qk_bias
+            query_states, key_states = torch.unbind(query_key, dim=2)
+
+            query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
+            key_states = key_states.view(bsz, q_len, self.num_heads, self.head_dim).transpose(1, 2)
+        else:
+            hidden_states = self.lf_gate(hidden_states, before_hidden_states)
+            query_states = self.q_proj(hidden_states)
+            key_states = self.k_proj(hidden_states)
+            qk_states = torch.cat([query_states, key_states], dim=-1)
+            qk_states = qk_states.view(bsz, q_len,
+                                    self.num_heads,
+                                    int(qk_states.shape[-1]//self.num_heads))
+            (query_states, key_states) = torch.chunk(qk_states, 2, dim=-1)
+            query_states = query_states.transpose(1, 2)
+            key_states = key_states.transpose(1, 2)
```

Comment:
these cat / transpose logic seems a little complex... Is it original implementation look likes this ?